### ISSUE TITLE
Singleton already set

### ISSUE THREAD

**Issue by omersezer** [roubachof/Sharpnado.Presentation.Forms#19]
when I put the application down and open again, it gives an error : singleton already set

**Comment by sander1133** [comment]
Same issue here. I'am using package "Sharpnado.Forms.HorizontalListView" version 0.9.8.
How to fix? Thank you in advance?

![1](https://user-images.githubusercontent.com/10980587/51787163-7dab2c80-216e-11e9-9e4e-2066a29f7301.png)

**Comment by roubachof** [comment]
It happens when you put the app in background and right after put it in foreground? Or you have to wait to be tombstoned then waken up by the system?

**Comment by sander1133** [comment]
It happens only when i use the Android back button to close the app and then try to open the App again. When just closing with the home button the error isn't occurring.

**Comment by roubachof** [comment]
Ok, I tried to reproduce it with no luck on pixel and nexus 5.

I could remove the check for singleton but it wouldn't solve the scenario your app is in.
Most of the time you use a framework to handle app Init (like registering services in your DI container, initializing ui libraries, etc...).

This library is responsible to ensure those services are initialized. It's also responsible to initialize one and only one those services. In your case it seems the code in charge of managing initialisation is doing it twice at a certain point of your app lifecycle.

You should change your code to see if your services are already initialized and not initialize them a second time.

You can check those two files in my Silly! app:

* https://github.com/roubachof/Xamarin-Forms-Practices/blob/master/SillyCompany.Mobile.Practices/CoreEntryPoint.cs
* https://github.com/roubachof/Xamarin-Forms-Practices/blob/master/SillyCompany.Mobile.Practices.Droid/DroidEntryPointSingleton.cs

You can see in the ```DroidEntryPointSingleton```, I keep a ```_isInitialized``` field to prevent situations like the one you describe to happen.

To be more clear with pseudo code:

```
class Initializer
{
    bool _isInitialized;

    Task InitializeAsync()
    {
        if (_isInitialized)
        {
            return;
        }
 
        PlatformHelper.InitializeSingleton(new AndroidPlatformHelper());
        Container.RegisterService<ISillyRepo, SillyRepo>();
        .....etc
    }
}

**Comment by sander1133** [comment]
Thank you for the investigation and your answer! 

I've watched the silly repository and i understand the purpose for checking if it's initialized before. However i am using multiple packages which require a initialize. Most of them stop working if i don’t initialize them the second time the app starts. Looks like the behave different than the Sharpnado package. Also most of the documentation that I’ve read don’t include the need of a “Re-initialization” check.

For now I’ve wrapped the init call in a try catch. It's not the most beautiful solution, but it works. Beneath is my current code for initialization from MainActivity/OnCreate.

`protected override void OnCreate(Bundle bundle)
{
	ToolbarResource = Resource.Layout.Toolbar;
	TabLayoutResource = Resource.Layout.Tabbar;

	base.OnCreate(bundle);

	Xamarin.Essentials.Platform.Init(this, bundle);
	Plugin.CurrentActivity.CrossCurrentActivity.Current.Init(this, bundle);
	Rg.Plugins.Popup.Popup.Init(this, bundle);
	ZXing.Net.Mobile.Forms.Android.Platform.Init();
	FFImageLoading.Forms.Platform.CachedImageRenderer.Init(true);
	CarouselView.FormsPlugin.Android.CarouselViewRenderer.Init();
	FAB.Droid.FloatingActionButtonRenderer.InitControl();
	Plugin.Iconize.Iconize.Init(Resource.Id.toolbar, Resource.Id.sliding_tabs);
	try
	{
		Sharpnado.Presentation.Forms.Droid.SharpnadoInitializer.Initialize();
	}
	catch (Exception) { }

	Forms.SetFlags("FastRenderers_Experimental");
	Forms.Init(this, bundle);

	LoadApplication(new App());
}`

**Comment by roubachof** [comment]
Should be resolved by 0.9.10.
